File: module-21-research-methodology-ctf-practice/research_notes_checker.py
"""
research_notes_checker.py

Validates a research log against the minimum structure a real security
research writeup (or a comprehensive exam research proposal) needs --
the same completeness-and-consistency checking pattern as Module 19's
capstone_checker.py and Module 09's VLAN plan validator, applied to a
RESEARCH log instead of a network build.

This does not judge whether your research idea is GOOD -- it checks
whether the standard structural pieces (a stated question, related work,
a methodology, an ethics/threat-model statement, results, and stated
limitations) are actually present, the same floor-not-ceiling framing
Module 19 used.

Usage:
    python research_notes_checker.py sample_research_log_complete.json
"""
import json
import sys


def check_research_question(log: dict) -> list:
    issues = []
    rq = log.get("research_question", "").strip()
    if not rq:
        issues.append("No research question stated at all.")
    elif len(rq.split()) < 5:
        issues.append(
            f"Research question ('{rq}') is too short to be a real, "
            f"specific question -- a comprehensive exam question needs "
            f"enough specificity to be falsifiable, not just a topic name."
        )
    if not rq.rstrip().endswith("?"):
        issues.append("Research question does not end in '?' -- state it "
                       "as an actual question, not a topic statement.")
    return issues


def check_related_work(log: dict) -> list:
    issues = []
    related = log.get("related_work", [])
    if len(related) < 2:
        issues.append(
            f"Only {len(related)} related work entr(y/ies) -- a real "
            f"literature review needs enough prior work cited to show "
            f"you know what's already been done, not just one paper."
        )
    for i, entry in enumerate(related):
        if not entry.get("citation"):
            issues.append(f"related_work[{i}] has no citation.")
        if not entry.get("relevance"):
            issues.append(
                f"related_work[{i}] ({entry.get('citation', '?')}) has no "
                f"stated relevance -- citing a paper without saying HOW it "
                f"relates to your question is not a literature review."
            )
    return issues


def check_methodology(log: dict) -> list:
    issues = []
    methodology = log.get("methodology", {})
    if not methodology.get("description"):
        issues.append("No methodology description -- how would someone "
                       "else reproduce this work?")
    if not methodology.get("ethics_statement"):
        issues.append(
            "No ethics/authorization statement -- for security research "
            "specifically, this is not optional (see Module 16's "
            "authorization-boundary theory, and this module's "
            "responsible-disclosure material)."
        )
    return issues


def check_results_and_limitations(log: dict) -> list:
    issues = []
    results = log.get("results", {})
    if not results.get("summary"):
        issues.append("No results summary.")
    if not results.get("evidence"):
        issues.append(
            "Results claim findings but cite no evidence -- an unsupported "
            "claim is not a finding, the same evidence-to-conclusion "
            "standard Module 18's forensic reconstruction required."
        )
    limitations = log.get("limitations", [])
    if not limitations:
        issues.append(
            "No stated limitations -- every real study has some; a "
            "research writeup with zero acknowledged limitations reads as "
            "either incomplete or overconfident, both of which a "
            "comprehensive exam committee will notice."
        )
    return issues


def run_checker(log: dict) -> dict:
    all_issues = []
    all_issues += check_research_question(log)
    all_issues += check_related_work(log)
    all_issues += check_methodology(log)
    all_issues += check_results_and_limitations(log)

    total_checks = 4
    checks_with_issues = len({
        "question" if any("question" in i.lower() for i in all_issues) else None,
        "related" if any("related work" in i.lower() or "related_work" in i.lower() for i in all_issues) else None,
        "methodology" if any("methodology" in i.lower() or "ethics" in i.lower() for i in all_issues) else None,
        "results" if any("results" in i.lower() or "limitation" in i.lower() for i in all_issues) else None,
    } - {None})
    completeness_pct = round(100 * (total_checks - checks_with_issues) / total_checks)

    return {"issues": all_issues, "completeness_pct": completeness_pct}


def main():
    path = sys.argv[1] if len(sys.argv) > 1 else "sample_research_log_complete.json"
    with open(path) as f:
        log = json.load(f)

    result = run_checker(log)
    print(f"Research notes checker -- {path}\n")
    print(f"Completeness: {result['completeness_pct']}%\n")

    if not result["issues"]:
        print("No gaps found. Every section is present and structurally sound.")
    else:
        print(f"{len(result['issues'])} gap(s) found:")
        for issue in result["issues"]:
            print(f"  - {issue}")


if __name__ == "__main__":
    main()

File: module-21-research-methodology-ctf-practice/research_notes_checker.js
/**
 * research_notes_checker.js
 *
 * Node port of research_notes_checker.py -- identical checks, identical
 * completeness scoring.
 *
 * Usage:
 *   node research_notes_checker.js sample_research_log_complete.json
 */
const fs = require("fs");

function checkResearchQuestion(log) {
  const issues = [];
  const rq = (log.research_question || "").trim();
  if (!rq) {
    issues.push("No research question stated at all.");
  } else if (rq.split(/\s+/).length < 5) {
    issues.push(
      `Research question ('${rq}') is too short to be a real, specific question -- a comprehensive exam question needs enough specificity to be falsifiable, not just a topic name.`
    );
  }
  if (!rq.trim().endsWith("?")) {
    issues.push(
      "Research question does not end in '?' -- state it as an actual question, not a topic statement."
    );
  }
  return issues;
}

function checkRelatedWork(log) {
  const issues = [];
  const related = log.related_work || [];
  if (related.length < 2) {
    issues.push(
      `Only ${related.length} related work entr(y/ies) -- a real literature review needs enough prior work cited to show you know what's already been done, not just one paper.`
    );
  }
  related.forEach((entry, i) => {
    if (!entry.citation) {
      issues.push(`related_work[${i}] has no citation.`);
    }
    if (!entry.relevance) {
      issues.push(
        `related_work[${i}] (${entry.citation || "?"}) has no stated relevance -- citing a paper without saying HOW it relates to your question is not a literature review.`
      );
    }
  });
  return issues;
}

function checkMethodology(log) {
  const issues = [];
  const methodology = log.methodology || {};
  if (!methodology.description) {
    issues.push("No methodology description -- how would someone else reproduce this work?");
  }
  if (!methodology.ethics_statement) {
    issues.push(
      "No ethics/authorization statement -- for security research specifically, this is not optional (see Module 16's authorization-boundary theory, and this module's responsible-disclosure material)."
    );
  }
  return issues;
}

function checkResultsAndLimitations(log) {
  const issues = [];
  const results = log.results || {};
  if (!results.summary) {
    issues.push("No results summary.");
  }
  if (!results.evidence) {
    issues.push(
      "Results claim findings but cite no evidence -- an unsupported claim is not a finding, the same evidence-to-conclusion standard Module 18's forensic reconstruction required."
    );
  }
  const limitations = log.limitations || [];
  if (limitations.length === 0) {
    issues.push(
      "No stated limitations -- every real study has some; a research writeup with zero acknowledged limitations reads as either incomplete or overconfident, both of which a comprehensive exam committee will notice."
    );
  }
  return issues;
}

function runChecker(log) {
  let allIssues = [];
  allIssues = allIssues.concat(checkResearchQuestion(log));
  allIssues = allIssues.concat(checkRelatedWork(log));
  allIssues = allIssues.concat(checkMethodology(log));
  allIssues = allIssues.concat(checkResultsAndLimitations(log));

  const totalChecks = 4;
  const categories = new Set();
  for (const i of allIssues) {
    const low = i.toLowerCase();
    if (low.includes("question")) categories.add("question");
    if (low.includes("related work") || low.includes("related_work")) categories.add("related");
    if (low.includes("methodology") || low.includes("ethics")) categories.add("methodology");
    if (low.includes("results") || low.includes("limitation")) categories.add("results");
  }
  const completenessPct = Math.round((100 * (totalChecks - categories.size)) / totalChecks);

  return { issues: allIssues, completeness_pct: completenessPct };
}

function main() {
  const filePath = process.argv[2] || "sample_research_log_complete.json";
  const log = JSON.parse(fs.readFileSync(filePath, "utf8"));

  const result = runChecker(log);
  console.log(`Research notes checker -- ${filePath}\n`);
  console.log(`Completeness: ${result.completeness_pct}%\n`);

  if (result.issues.length === 0) {
    console.log("No gaps found. Every section is present and structurally sound.");
  } else {
    console.log(`${result.issues.length} gap(s) found:`);
    for (const issue of result.issues) {
      console.log(`  - ${issue}`);
    }
  }
}

main();
